### DISCUSSION TITLE
TPI 2 - Operaciones asincrónicas

### DISCUSSION THREAD

**Discussion by agmblanco** [dds-jv/foro#428] [TPI3 - Sistema de Archivos] [answered]
Buenas noches profes. Tengo una duda respecto de las operaciones asincrónicas. Inicialmente había contemplado una abstracción que me permitiese llevar un registro de las operaciones de lectura y escritura asincrónicas pendientes, con el objetivo de, al momento de que se quiera cerrar el archivo, poder devolver una excepción indicando que existen operaciones pendientes en el mismo y por ende, no se puede cerrar. Pero lo estuve reconsiderando ya que no sé si eso sería extender el dominio. 
En particular, el fragmento de código que me hace dudar es:
```java
  public void cerrar(){
    if(!abierto){
      throw new RuntimeException("El archivo ya fue cerrado");
    }
    if(!operacionesPendientes.isEmpty()){
      throw new RuntimeException("El archivo se encuentra ocupado. No se puede cerrar hasta que se completen las operaciones pendientes");
    }
    this.sistemaBase.closeFile(this.descriptor);
  }
```
¿Tengo permitido hacer ese tipo de verificaciones o estaría incumpliendo los requerimiento? 

Desde ya muchas gracias. Quedo a la espera de su respuesta, Agustín Blanco.

**Comment by flbulgarelli** [comment, marked as answer]
:wave: ¡Hola! 

No lo pretendíamos, pero definitivamente es una buena idea. Lo que no me queda claro es cómo harás para saber qué operaciones quedan pendientes; si bien llevar esa contabilidad es posible a través de envolver las operaciones asincrónicas del código cliente dentro de otras provistas por tu adaptador que se encarguen de registrar las operaciones incompletas, hacerlo puede ser complejo (sobre todo, sin caer en _condiciones de carrera_).

**Reply by agmblanco** [reply]
Hola Franco, muchas gracias por la respuesta! 

Finalmente decidí implementarlo. Para saber qué operaciones quedan pendientes encapsulé los respectivos callbacks en callbacks propios que, además de ejecutar lo esperado por el usuario, se ocupasen de mantener actualizado las operaciones pendientes del archivo. 

Por ejemplo, en el caso de la escritura asincrónica, al registrarse el callback provisto por el usuario, éste se guarda como parámetro dentro de una instancia de `EscrituraPendiente`, clase que implementa la interfaz `OperacionPendiente`. Y, al registrar el callback en el `LowLevelFileSystem`, en vez de registrar el provisto por el usuario, registro un método de mi objeto instanciado.

Archivo.java
```java
  public void finalizarOperacionAsincronica(OperacionAsincronica operacion) {
    if (!this.operacionesPendientes.contains(operacion)) {
      throw new OperacionInvalidaException("La operación no existe o no pertenece a este archivo");
    }
    this.operacionesPendientes.remove(operacion);
  }
  
  public void escribirAsincronico(Buffer buffer, Runnable callback) {
    EscrituraPendiente escrituraPendiente = new EscrituraPendiente(this, callback);
    this.operacionesPendientes.add(escrituraPendiente);
    sistemaBase.asyncWriteFile(
        descriptor,
        buffer.obtenerDatos(),
        buffer.obtenerInicio(),
        buffer.obtenerFin(),
        escrituraPendiente::finalizar
    );
  }
```
EscrituraPendiente.java
```java
package fs;

import java.util.function.Consumer;

public class EscrituraPendiente implements OperacionAsincronica {
  private Archivo archivo;
  private Runnable callback;

  public EscrituraPendiente(Archivo archivo, Runnable callback) {
    this.archivo = archivo;
    this.callback = callback;
  }

  public void finalizar() {
    callback.run();
    archivo.finalizarOperacionAsincronica(this);
  }
}
```

En cuanto a las condiciones de carrera, lo estuve pensando un buen rato junto con un compañero y realmente no logramos encontrar ninguna. Obviamente esto no quita que puedan existir, pero ciertamente no fueron detectadas.

Quedo atento a las correcciones. Saludos cordiales, Agustín M. Blanco.
